### ISSUE TITLE
Possible threading issues with non thread safe collections in Options classes

### ISSUE THREAD

**Issue by ohads-MSFT** [dotnet/dotnet-api-docs#8658]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Options use caching, for example:
https://github.com/dotnet/runtime/blob/e2c0735927642a11162c0d12799be2921bd1c984/src/libraries/Microsoft.Extensions.Options/src/UnnamedOptionsManager.cs#L25

Since [`TOptions<TOptions>` are registered at the singleton scope](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/configuration/options?view=aspnetcore-7.0#options-interfaces), any two DI participants (e.g. controller instances) may receive the same `TOptions` instance when they retrieve it using `IOptions<T>.Value`.

This means that any user of the Options class must assume concurrent access to the classes' members. Now, for most boundable types (like `string` or `int`) this won't be an issue. However, for collection types, this potentially becomes more complicated. Even if the first thing I do is store a `IOptions<T>.Value.ToImmutableArray()` , this still incurs enumeration. It is not clear to me that all boundable collection types are thread safe even for read-only enumeration of the original collection in the shared `TOptions` class. And while I suspect [arrays possess this property](https://learn.microsoft.com/en-us/dotnet/api/system.array?view=net-7.0#thread-safety), I do not see any such mention for e.g. [`HashSet<T>`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1?view=net-7.0).

### Expected Behavior

Guarantee at the ASP.NET level that all boundable Options collection types are thread safe for concurrent enumeration. For example, if `HashSet<T>` does not provide this guarantee, you could create a subclass of it that does provide it, and inject it into bound `TOptions` classes.

### Steps To Reproduce

```cs
public class MyOptions { HashSet<string> MySet {get; set;} }

public class MyController
{
   public MyController(IOptions<MyOptions> options)
   {
      foreach (string s in options.Value.MySet) { Console.WriteLine(s);  }
   }
}
```
### Exceptions (if any)
In theory, things like `NullReferenceException` or similar unexpected exceptions when non thread safe operations are performed.

### .NET Version
6.0.403

### Anything else?
In case I'm wrong and all boundable collections are thread safe for readonly enumeration, I believe it should be specified clearly in all the relevant documentation pages. I guess in that case that would be more of a dotnet docs bug, but considering the aforementioned singleton registration of `IOptions<T>` I think it's something you should address in your docs too.

**Comment by eiriktsarpalis** [comment]
> It is not clear to me that all boundable collection types are thread safe even for read-only enumeration of the original collection in the shared TOptions class. And while I suspect [arrays possess this property](https://learn.microsoft.com/en-us/dotnet/api/system.array?view=net-7.0#thread-safety), I do not see any such mention for e.g. [HashSet<T>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1?view=net-7.0).

Moving to dotnet-api-docs specifically to track this concern. We don't explicitly document that read-only access of collections is thread-safe, as we mostly regard this as true by definition. Even though [it is possible](https://ericlippert.com/2011/05/23/read-only-and-threadsafe-are-different/) for collection implementations to break that invariant, our implementations in System.Collections don't do this (or if they do this accidentally, we will treat is a bug).

**Comment by stephentoub** [comment]
> @stephentoub thoughts? Do you think such a clarification is warranted?

I think it'd be good to document on a collection-by-collection basis that they're thread-safe for concurrent reads (and nothing else, e.g. they're not safe to read and write concurrently, not safe to write and write concurrently, etc.).  We certainly could never change that for our core types like List, Dictionary, and HashSet, as loads of code (including our own) depends on that.

**Comment by ohads-MSFT** [comment]
> We don't explicitly document that read-only access of collections is thread-safe, as we mostly regard this as being true by definition. 

This is further complicated by the fact that sometimes it IS documented, for example in [List<T> docs](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1?view=net-7.0):
> It is safe to perform multiple read operations on a [List<T>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1?view=net-7.0)

Since the [HashSet<T> docs](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1?view=net-7.0) don't mention thread safety at all, the reasonable consumer assuming MS docs are consistent could (incorrectly) deduce that this does not hold for `HashSet<T>`.

In addition, the [thread-safety section of arrays](https://learn.microsoft.com/en-us/dotnet/api/system.array?view=net-7.0#thread-safety) is confusing at best, and misleading at worst:
> Enumerating through a collection is intrinsically not a thread-safe procedure
For me it's not clear what this actually means, and specifically IMHO the wording doesn't make it sound like the invariant discussed here actually holds (safe for readonly)...

**Comment by eiriktsarpalis** [comment]
> > Enumerating through a collection is intrinsically not a thread-safe procedure
> 
> For me it's not clear what this actually means, and specifically IMHO the wording doesn't make it sound like the invariant discussed here necessarily holds (safe for readonly)...

I'd say that's clarified by the following sentences, full excerpt:

> Enumerating through a collection is intrinsically not a thread-safe procedure. Even when a collection is synchronized, other threads can still modify the collection, which causes the enumerator to throw an exception. To guarantee thread safety during enumeration, you can either lock the collection during the entire enumeration or catch the exceptions resulting from changes made by other threads.

**Comment by stephentoub** [comment]
Even without invalidating a version number, reading an array while someone else writes to it may not be "thread safe", e.g. if elements in the array aren't read/written atomically and may be torn.

**Comment by ohads-MSFT** [comment]
> I'd say that's clarified by the following sentences
I agree it is implied, but IMO the wording is still confusing, and it does not explicitly state that pure readonly scenarios are safe. It's also different than the wording used in `List<T>` so again, an issue of consistency.

**Comment by GSPP** [comment]
Maybe it can be pointed out that an array will behave as a set of independent memory locations. It's just like declaring a set of fields (`int item0, item1, item2, ...`). From that statement, essentially all the thread-safety properties of arrays follow automatically.

**Comment by GSPP** [comment]
An attempt for `Array`:

> In terms of thread-safety, the elements of an array behave as a set of independent memory locations. Thread-safety can therefore be understood in terms of the read and write operations performed on each element separately.
>
> Enumeration is carried out be sequentially reading all elements individually from lowest to highest index. No particular synchronization is performed. In the presence of concurrent writes, this can lead to various concurrency effects which may or may not be acceptable. Concurrent reads, including other enumeration operations, have no effect.
